share/eds/store.go

```diff
@@ -197,31 +201,59 @@ func (s *Store) Blockstore() bstore.Blockstore {
 	return s.bs
 }
 
-// CARBlockstore returns the IPFS blockstore that provides access to the IPLD blocks stored in an
-// individual CAR file.
-func (s *Store) CARBlockstore(ctx context.Context, dataHash []byte) (dagstore.ReadBlockstore, error) {
-	key := shard.KeyFromString(fmt.Sprintf("%X", dataHash))
+// CARBlockstore returns an IPFS Blockstore providing access to individual shares/nodes of a
+// specific EDS identified by DataHash and registered on the Store. NOTE: The Blockstore does not
+// store whole Celestia Blocks but IPFS blocks. We represent `shares` and NMT Merkle proofs as IPFS
+// blocks and IPLD nodes so Bitswap can access those.
+func (s *Store) CARBlockstore(
+	ctx context.Context,
+	root share.DataHash,
+) (dagstore.ReadBlockstore, error) {
+	key := shard.KeyFromString(root.String())
 	accessor, err := s.getAccessor(ctx, key)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("eds/store: failed to get accessor: %w", err)
 	}
-
 	return accessor.bs, nil
 }
 
-func (s *Store) getAccessor(ctx context.Context, key shard.Key) (*accessorWithBlockstore, error) {
-	// try to fetch from cache
-	accessor, err := s.cache.Get(key)
-	if err != nil && err != errCacheMiss {
-		log.Errorw("unexpected error while reading key from bs cache %s: %s", key, err)
+// GetDAH returns the DataAvailabilityHeader for the EDS identified by DataHash.
+func (s *Store) GetDAH(ctx context.Context, root share.DataHash) (*share.Root, error) {
+	key := shard.KeyFromString(root.String())
+	accessor, err := s.acquireShard(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("eds/store: failed to get accessor: %w", err)
 	}
-	if accessor != nil {
-		return accessor, nil
+	defer accessor.Close()
+
+	reader, err := carv1.NewCarReader(accessor)
+	if err != nil {
+		return nil, fmt.Errorf("eds/store: failed to read DAH from car reader: %w", err)
 	}
 
-	// wasn't found in cache, so acquire it and add to cache
+	dah := dahFromCARHeader(reader.Header)
+	if !bytes.Equal(dah.Hash(), root) {
```

Comment:
I wonder if we should handle this one a bit more seriously, then just returning an error to caller. It is strong sign of eds store integrity corruption.

Reply:
We can utilize DagStores recovery mechanisms, etc. Though we can keep it for later. At this point, we just trust SSDs a node is setup on :)

Reply:
Content integrity mismatch is handed up in a similar way from the `dagstore`, so I think I'll keep it like this. But I'm on the fence about this.. panic also is justified here

Reply:
This is not necessarily Disk corruption, but could be data being overwritten by third party, or dagstore bug, or data was corrupted before it has been stored.
The bad thing about this one is that it will be thrown again with same request. It definitely should notify node operator if no fallback recovery logic is triggered.
Anyway, I think this should be discussed/solved outside of the scope of this PR, so resolving for now.